theseus/optimizer/nonlinear/nonlinear_least_squares.py

```diff
@@ -32,21 +72,329 @@ def __init__(
         step_size: float = 1.0,
         **kwargs,
     ):
-        linear_solver_cls = linear_solver_cls or CholeskyDenseSolver
         super().__init__(
             objective,
             linear_solver_cls=linear_solver_cls,
             vectorize=vectorize,
-            linearization_cls=linearization_cls,
-            linearization_kwargs=linearization_kwargs,
-            linear_solver_kwargs=linear_solver_kwargs,
             abs_err_tolerance=abs_err_tolerance,
             rel_err_tolerance=rel_err_tolerance,
             max_iterations=max_iterations,
             step_size=step_size,
             **kwargs,
         )
+        linear_solver_cls = linear_solver_cls or CholeskyDenseSolver
```

Comment:
`NonlinearLeastSquares` still seems to be making assumption that solver/optimizer is using linearization to solve. There might be `Nonlinearleastsquare` optimizers not using linearization. I see that it's being done so that subclasses such as `gauss-newton` and `LM` can be derived from this. But I feel there should one more level of abstraction between NonlinearLeastSquares and GN/LM optimizer class.

Reply:
This is a reasonable point, but I wonder if it's worth it to preemptively introduce a more complicated hierarchy for an optimizer that we don't currently have. At least for DCEM, what I'm picturing is for it to be more general than NLLS via the `error_metric` function, so it should inherit directly from `NonlinearOptimizer`. But maybe GBP? 

@joeaortiz  and @mhmukadam, is GBP meant to be used only for NLLS? If that's the case then it makes sense to do 

`NonlinearLeastSquares` --> `GBP` (NLLS only defines the error metric)
`NonlinearLeastSquares --> SomeCoolName --> GN/LM` (most of the logic would go to the intermediate class)

But otherwise, seems cleaner o leave as is until we have some optimizer for NLLS that doesn't use the type of linearization that GN and LM uses.

Reply:
Since these are optimizers, in my mind `NonlinearLeastSquares` are methods that linearize the objective and call a linear solver iteratively. The confusion might be thinking about the name as solving nonlinear least squares objectives.

Luis is correct that CEM should derive from `NonlinearOptimizer` even if it only supports nonlinear least squares objectives right now, since the objective type it could support is more general.

GBP seems tricky since it does involve local linearization and local linear solves. Since this PR is blocking CEM progress let's proceed assuming `NonlinearOptimizer` --> `GBP` and then we can make any updates in the GBP PR once @joeaortiz has more bandwidth.

Reply:
Agree on doing `NonlinearOptimizer --> GBP` for now.

Also, IIRC, CEM should already very easily support other objectives, by replacing all calls current calls to error squared norm with `error_metric()`. 

That being said, and also relevant to the discussion above, it seems to me that the `error_metric()` method more appropriately belongs to `Objective` class. Otherwise, we will need to have some nonlinear optimizer subclass just so that users can pass the error metric to use, which seems strange.

Reply:
Yeah makes sense that error metric is a function of the objective. But do we need to make some class hierarchy for objective with the common stuff in a base class `Objective` and the NLS related stuff in a `NLSObjective`?

Reply:
Maybe, but I'd rather not go too crazy right now. We can add a subclass that just hardcodes the squared error metric, but leave all of the other stuff inside the base class. And revisit this if needed.

Reply:
Agreed not needed yet or in this PR.

So have we settled on merging after moving error_metric to objective or do we need anything else? cc @dishank-b

Reply:
Yeah, LGTM! We can merge.

Reply:
Made #455 for this feature. If that looks good, I'll merge both PRs.

Reply:
One more thing, I will still be using the very similar `_optimize_impl` and `_optimize_loop` core structure for DCem, the different is instead of using linearization based solving i would use DCem solver. From that perspective, would it make sense to have more general `_optimize_impl` and `_optimize_loop` in `NonlinearOptimizer`? 
I can also just copy from here, but that would make it repetitive.

Reply:
I'm less clear on what a general version of these two methods would look like. Let's start by merging this PR, updating your PR with these changes, and possibly have a second round of refactoring after some review. I'd like to see what exactly is the diff between your version of `_optimize_loop` and `_optimize_impl` after we merge these changes.